### ISSUE TITLE
hierarchical_topics() produce incorrect output when three topics have the same distance

### ISSUE THREAD

**Issue by azikoss** [MaartenGr/BERTopic#1907]
Hi there,

I have noticed that `hierarchical_topics(...)` method produces incorrect results when three or more topics have the same (tf-idf) distances. Let me illustrative it with an example.

```
from umap import UMAP
from bertopic import BERTopic

docs = (
    ["banana"] * 300 
    + ["banana apple"] * 300 
    + ["pear"] * 300 
    + ["lemon"] * 300 
    + ["clock"] * 300 
)

model = BERTopic(umap_model=UMAP(random_state=42))
topics, probs = model.fit_transform(docs)
hr = model.hierarchical_topics(docs)
hr
```
This outputs
![image](https://github.com/MaartenGr/BERTopic/assets/124801530/c403ae15-e974-48dd-83eb-6842cf2b7d6a)

The cluster with Parent_ID == 8 includes topics [1, 2, 3] but topic 3 is not mentioned in either the left or right child or their childs. 

**Why is it happening?** 
The [flat structure](https://github.com/MaartenGr/BERTopic/blob/master/bertopic/_bertopic.py#L1006) is created in each iteration. There is no guarantee that the new cluster will contain only two topics while the [code that follows](https://github.com/MaartenGr/BERTopic/blob/master/bertopic/_bertopic.py#L1031-L1041) presumes that.

**What should be the expected behavior?**
I think that a new cluster should emerge. Essentially, the structure should look like this

```
Parent_Id, Child_Left_ID, Child_Right_ID
     8               1               11
    11               2               3
```

**Comment by MaartenGr** [comment]
Thank you for the reproducible example! This indeed seems to be an issue. Interestingly, I expect this would not happen often since distances are seldom the same right? How did you then stumble upon this issue? 

Either way, seeing as you already dove into the code, any suggestions how this could easily be resolved? Perhaps check for similar distances beforehand?

**Comment by MaartenGr** [comment]
> A possible fix could be (deterministically) select a topic that has the same distance ith two other topics and add a tiny bit to the distance metric. 

If it's small enough that it's not significant, this would be the easiest solution without needing to change much to the source code. We could also add some (very slight perhaps uniform) noise to the entire distance metric such that all distances are affected equally. Either way, this sounds like a good approach as long as we can keep the noise small and logical. 

If you have the time, a PR would be appreciated! If not, I'll put it on my backlog but I'm not sure if I'm able to tackle this in the coming months.

**Comment by azikoss** [comment]
I like the approach that adds the noise uniformly! Let me work it out.
